### DISCUSSION TITLE
In blink.cmp, how do i put AI result on top?

### DISCUSSION THREAD

**Discussion by ogios** [LazyVim/LazyVim#5093] [Q&A] [answered]
I'm using codeium.

the AI result will be updated and put on top of the completion list in cmp, but it does not do that since i've updated the lazyvim and use the blink.cmp.

this hurts when i want to use AI completion, and i don't even know if AI responded or not because i have to scroll through the list to find it

**Comment by ogios** [comment]
it would seem that we can only do this with opts.fuzzy.sorts, but that's a hard work i think?

**Reply by dpetka2001** [reply]
You could ask in `blink` Discussions and maybe the maintainer can give you an example of how you could do that and you can work your way from there. Or just re-enable `nvim-cmp` from LazyExtras if it aligns more with your personal preference.

**Reply by ogios** [reply]
@dpetka2001 thanks, i found one way to do that, but i encountered a problem: https://github.com/LazyVim/LazyVim/discussions/5093#discussioncomment-11572290

**Comment by ogios** [comment]
i'm frustrated
```lua
          function(a, b)
            -- return a.score > b.score -- this works
            return a.source_id == "codeium" -- this will raise: Error executing vim.schedule lua callback: ...l/share/nvim/lazy/blink.cmp/lua/blink/cmp/fuzzy/sort.lua:11: invalid order function for sorting
          end,
```

why??? they all returned `true` or `false` right?

inside it uses this:
```lua
function sort.sort(list)
  local config = require('blink.cmp.config').fuzzy.sorts
  local sorting_funcs = vim.tbl_map(
    function(name_or_func) return type(name_or_func) == 'string' and sort[name_or_func] or name_or_func end,
    config
  )
  table.sort(list, function(a, b)
    for _, sorting_func in ipairs(sorting_funcs) do
      local result = sorting_func(a, b)
      if result ~= nil then return result end
    end
  end)
  return list
end
```

**Reply by minusfive** [reply]
You may want to ask in the `blink.cmp` repo.

**Reply by ogios** [reply]
> You may want to ask in the `blink.cmp` repo.

yeah, just i thought it's some special thing with lua, cause i don't see any blink specific logic in that sort code... the code is rather simple but i just can't understand why...

**Reply by dpetka2001** [reply]
Can you try
```lua
function(a, b) 
-- return a.score > b.score
if a.source_id == "codeium" then
  return true
end
if b.source_id == "codeium" then
  return false
end
```
And see if that works?

**Reply by ogios** [reply]
@dpetka2001 thanks!
but same error:
```
Error executing vim.schedule lua callback: ...l/share/nvim/lazy/blink.cmp/lua/blink/cmp/fuzzy/sort.lua:11: invalid order function for sorting
stack traceback:
	[C]: in function 'sort'
	...l/share/nvim/lazy/blink.cmp/lua/blink/cmp/fuzzy/sort.lua:11: in function 'fuzzy'
	...re/nvim/lazy/blink.cmp/lua/blink/cmp/completion/list.lua:109: in function 'fuzzy'
	...re/nvim/lazy/blink.cmp/lua/blink/cmp/completion/list.lua:85: in function 'show'
	...re/nvim/lazy/blink.cmp/lua/blink/cmp/completion/init.lua:34: in function <...re/nvim/lazy/blink.cmp/lua/blink/cmp/completion/init.lua:29>
```

**Reply by dpetka2001** [reply]
Not at home right now and on mobile. Will try to take a look later tonight. In the mean time you could also create a Discussion in blink and ask there. I'm sure the maintainer will guide you. I asked something about disabling `cmdline` before Lazyvim made the change and he was really helpful.

**Reply by ogios** [reply]
@dpetka2001 thanks, i've posted a discussion there

**Reply by dpetka2001** [reply]
@ogios So, i just tried to emulate the logic that Blink uses in its sorting functions and ended up with the following
```lua
fuzzy = {
        -- sorts = { prioritize_codeium },
        sorts = {
          function(a, b)
            -- Check if either item is from Codeium
            local a_is_codeium = a.source_id == "codeium"
            local b_is_codeium = b.source_id == "codeium"

            -- If both or neither are from Codeium, let other sorters decide
            if a_is_codeium == b_is_codeium then
              return
            end

            -- Return true to place Codeium items first
            return a_is_codeium
          end,
        },
      },
```
Please test it out.

**Reply by dpetka2001** [reply]
Even your initial solution works with this modification
```lua
function(a, b)
-- return a.score > b.score -- this works
  if a.source_id == b.source_id then
    return
  end
  return a.source_id == "codeium" -- This doesn't raise error any more
end,
```
It seems the key is to also define what to do in case the 2 entries are the same (return nil), so that other sorters down the list will decide what to do I guess??

**Reply by folke** [reply, marked as answer]
fyi, I just pushed some updates related to blink and ai providers.
AI providers now get a score boost of `100` and I've also enabled `async=true`, which means
the blink menu will already show while ai results are still fetched.

**Reply by dpetka2001** [reply]
Great!! Thank you @folke as always :smile:

**Reply by folke** [reply]
I'm using `vim.g.cmp_ai = false` in my config, so I never noticed that slow-down.
When I just tested, I was very very surprised to say the least.
Luckily adding `async = true` makes completions instant :)

**Reply by ogios** [reply]
@dpetka2001 thanks.
i think it might related to the table.sort logic, i only use the first argument to judge ture or false, so if:
(a, b) -> ture and next time (b, c) -> ture and then (c, a) -> true

this is wrong, cause a>b and b>c, so a must bigger than c but the function return true which means c>a, the sort function don't know what to do and raise the error.

but i can not be sure of it's internal logic
